cmake/BuildInfo.cmake

```diff
@@ -0,0 +1,112 @@
+cmake_minimum_required(VERSION 3.14)
+
+# These values indicate "something went wrong"
+set(PRODUCT_BRANCH_ID   "0" CACHE STRING "Branch ID")
+set(PRODUCT_BUILD_ID    "0" CACHE STRING "Build ID")
+set(PRODUCT_BUILD_TYPE  "0" CACHE STRING "Build Type")
+
+if(PRODUCT_EMBED_BUILD_INFO)
+    find_package(Git QUIET)
+    if(Git_FOUND)
+        execute_process(
+            COMMAND ${GIT_EXECUTABLE} log --pretty=format:%H -n 1
+            OUTPUT_VARIABLE GIT_REV
+            ERROR_QUIET
+        )
+        execute_process(
+            COMMAND git diff --exit-code
+            OUTPUT_VARIABLE GIT_DIFF_CONTENTS
+            RESULT_VARIABLE GIT_DIFF
+            OUTPUT_STRIP_TRAILING_WHITESPACE
+            ERROR_QUIET
+        )
+        execute_process(
+            COMMAND git describe --exact-match --tags
+            OUTPUT_VARIABLE GIT_TAG
+            OUTPUT_STRIP_TRAILING_WHITESPACE
+            ERROR_QUIET
+        )
+        string(REGEX REPLACE "[\r\n]" " " GIT_TAG "${GIT_TAG}")
+
+        if(NOT GIT_TAG)
+            execute_process(
+                COMMAND ${GIT_EXECUTABLE} log --pretty=format:%h -n 1
+                OUTPUT_VARIABLE GIT_TAG
+                OUTPUT_STRIP_TRAILING_WHITESPACE
+                ERROR_QUIET
+            )
+        endif()
+        if(GIT_DIFF EQUAL 1)
+            string(APPEND GIT_REV "_dirty")
+            string(APPEND GIT_TAG "_dirty")
+        endif()
```

Comment:
Why not just use `git describe --tags --dirty`, which automatically appends `-dirty` when the repo is dirty?

Obviously we'd need to put some more reasonable tags in the repo...

Reply:
I remember bypassing `git describe` because it was using reaaaaaly old tags. We need to do a better job of tagging, heh. I'm not sure how I feel about seeing `UruLive.1.918 - gamemgr-1437-g87f077d1c - Internal.Release` over `UruLive.1.918 - g87f077d1c - Internal.Release`. In particular, I chose the `_dirty` prefix over `-dirty` to avoid confusion with the other dash separators in the product string.

Reply:
OK, I think I've got a workable solution such that we don't do this whole manual dirty thing and obsolete tags aren't picked up.